tempesta_fw/ss_skb.c

```diff
@@ -788,18 +788,13 @@ __skb_fragment(struct sk_buff *skb_head, struct sk_buff *skb, char *pspt,
 		d_size = skb_frag_size(frag);
 		offset = pspt - (char *)skb_frag_address(frag);
 
-		if (offset >= 0 && offset <= d_size) {
+		/*
+		 * @pspt can be the end of a frag, but it can also be a start of
+		 * another fragment. Not necessary, that both frags are
+		 * neighbours.
+		 */
+		if (offset >= 0 && offset < d_size) {
 			int t_size = d_size - offset;
-			if (!t_size) {
-				/*
-				 * @pspt is at the end of the frag (zero tail
-				 * length): append if @len > 0 or move to the
-				 * next frag for deletion.
-				 */
-				if (len > 0)
-					goto append;
-				continue;
-			}
```

Comment:
Very subtle place and `__split_pgfrag_add()` also optimizes zero offset, so quite a good fix.

Reply:
Yeah, we happily jumped over this bug. I was very lucky to catch it. Data corruption is possible on large response headers, but too many conditions must met. Happily this cannot be exploited by an attacker.